mongo/crud_examples_test.go

```diff
@@ -323,13 +323,20 @@ func ExampleCollection_Distinct() {
 	// run on the server.
 	filter := bson.D{{"age", bson.D{{"$gt", 25}}}}
 	opts := options.Distinct().SetMaxTime(2 * time.Second)
-	values, err := coll.Distinct(context.TODO(), "name", filter, opts)
+	cur, err := coll.Distinct(context.TODO(), "name", filter, opts)
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	for _, value := range values {
-		fmt.Println(value)
+	// Get a list of all returned documents and print them out.
+	// See the mongo.Cursor documentation for more examples of using cursors.
+	var results []bson.M
+	if err = cur.All(context.TODO(), &results); err != nil {
+		log.Fatal(err)
+	}
```

Comment:
I'm concerned that the `Cursor` API doesn't really fit the requirements of the `distinct` API. In the example code here, the user must provide a `Context` to iterate the cursor because the `Cursor` might have to call `getMore`. However, `distinct` returns all results and there is no server-side cursor to iterate, so that `Context` and `getMore` logic are unnecessary. Additionally, there's no way to access the full BSON result, only one value at a time.

Looking for uses of `Distinct` using [Sourcegraph](https://sourcegraph.com/search?q=context%3Aglobal+lang%3AGo++-repo%3A%5Egithub%5C.com%2Fmongodb%2Fmongo-go-driver%24+repo%3Ahas.content%28go.mongodb.org%2Fmongo-driver%29+-file%3A%28%5E%7C%2F%29vendor%2F+.Distinct%28c&patternType=standard&sm=1&groupBy=repo), I found that the most common code is something like:
```go
results, err := coll.Distinct(ctx, "myField", filter)
if err != nil {
	return err
}

reportTimes := make([]int64, len(results))
for i, result := range results {
	reportTimes[i] = result.(int64)
}
```
Users tend to expect the returned values to be all the same type. Handling of unexpected types varies.

A better user experience uses an API similar to `SingleResult`:
```go
var res []string
err := coll.Distinct(ctx, "myField", filter).Decode(&res)
if err != nil {
	return err
}
```

If a user wanted to handle different types with different logic, they could call `Raw` instead:
```go
res, err := coll.Distinct(ctx, "myField", filter).Raw()
if err != nil {
	return err
}

values, err := res.Values()
if err != nil {
	return err
}

for _, val := range values {
	switch val.Type {
	// Handle different values.
	}
}
```

It's not clear if the actual `SingleResult` type can be reused for `Distinct`, but at least the API seems more appropriate.